stdlib/public/Darwin/Foundation/NSDictionary.swift

```diff
@@ -77,49 +149,85 @@ extension Dictionary : _ObjectiveCBridgeable {
       return
     }
 
-    if Key.self == String.self {
+    let keyStride = MemoryLayout<Key>.stride
+    let valueStride = MemoryLayout<Value>.stride
+    let objectStride = MemoryLayout<AnyObject>.stride
+
+    if keyStride < objectStride || valueStride < objectStride {
+      // `Dictionary<Key, Value>` where either `Key` or `Value` is a value type
+      // may not be backed by an NSDictionary.
+      var builder = _DictionaryBuilder<Key, Value>(count: d.count)
+      d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
+        let anyObjectKey = anyKey as AnyObject
+        let anyObjectValue = anyValue as AnyObject
+        builder.add(
+            key: Swift._forceBridgeFromObjectiveC(anyObjectKey, Key.self),
+            value: Swift._forceBridgeFromObjectiveC(anyObjectValue, Value.self))
+      })
+      result = builder.take()
+    } else {
+      defer { _fixLifetime(d) }
+    
+      let numElems = d.count
+      
       // String and NSString have different concepts of equality, so
       // string-keyed NSDictionaries may generate key collisions when bridged
       // over to Swift. See rdar://problem/35995647
-      var dict = Dictionary(minimumCapacity: d.count)
-      d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
-        let key = Swift._forceBridgeFromObjectiveC(
-          anyKey as AnyObject, Key.self)
-        let value = Swift._forceBridgeFromObjectiveC(
-          anyValue as AnyObject, Value.self)
-        // FIXME: Log a warning if `dict` already had a value for `key`
-        dict[key] = value
-      })
-      result = dict
-      return
-    }
+      let needToHandleDuplicates = (Key.self == String.self)
+      
+      result = Dictionary(_unsafeUninitializedCapacity: numElems, allowingDuplicates: needToHandleDuplicates) { (keyBuffer, valueBuffer, initializedCount) in
+        
+        let anyKeyPtr = UnsafeMutableRawPointer(mutating: keyBuffer.baseAddress!).bindMemory(to: AnyObject.self, capacity: numElems)
+        let anyValuePtr = UnsafeMutableRawPointer(mutating: valueBuffer.baseAddress!).bindMemory(to: AnyObject.self, capacity: numElems)
 
-    // `Dictionary<Key, Value>` where either `Key` or `Value` is a value type
-    // may not be backed by an NSDictionary.
-    var builder = _DictionaryBuilder<Key, Value>(count: d.count)
-    d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
-      let anyObjectKey = anyKey as AnyObject
-      let anyObjectValue = anyValue as AnyObject
-      builder.add(
-          key: Swift._forceBridgeFromObjectiveC(anyObjectKey, Key.self),
-          value: Swift._forceBridgeFromObjectiveC(anyObjectValue, Value.self))
-    })
-    result = builder.take()
-  }
+        __NSDictionaryGetObjects(d, anyValuePtr, anyKeyPtr, numElems)
```

Comment:
It feels a little weird that `getObjects:andKeys:count:` fills the buffers with +0 references, but `UnsafeMutablePointer<AnyObject>` assumes +1. It seems fine though -- loading elements through the UMP will still do the right thing.

Elsewhere we use `AutoreleasingUnsafeMutablePointer` to make the semantics explicit. I'm not sure if it would help here, though -- it may just obfuscate the code.

Reply:
Arguably you can use `UnsafePointer<AnyObject>`, since you are just reading from these and that assumes no particular ownership semantics.